src/pip/_internal/download.py

```diff
@@ -142,31 +147,79 @@ def __init__(self, prompting=True):
         self.prompting = prompting
         self.passwords = {}
 
-    def __call__(self, req):
-        parsed = urllib_parse.urlparse(req.url)
+    def _get_url_and_credentials(self, original_url,
+                                 allow_netrc=True, allow_keyring=True):
+        """Returns the credentials to use for the provided URL.
+
+        If allowed, netrc and keyring may be used to obtain the
+        correct credentials.
+
+        Returns (url_without_credentials, username, password). Note
+        that even if the original URL contains credentials, the result
+        of this function may include different values.
+        """
+        parsed = urllib_parse.urlparse(original_url)
 
         # Split the credentials from the netloc.
         netloc, url_user_password = split_auth_from_netloc(parsed.netloc)
 
-        # Set the url of the request to the url without any credentials
-        req.url = urllib_parse.urlunparse(parsed[:1] + (netloc,) + parsed[2:])
+        # Return the url without any credentials
+        url = urllib_parse.urlunparse(parsed[:1] + (netloc,) + parsed[2:])
 
         # Use any stored credentials that we have for this netloc
         username, password = self.passwords.get(netloc, (None, None))
 
         # Use the credentials embedded in the url if we have none stored
         if username is None:
             username, password = url_user_password
+            if username or password:
+                logger.debug("Found credentials in url for %s", netloc)
 
         # Get creds from netrc if we still don't have them
-        if username is None and password is None:
-            netrc_auth = get_netrc_auth(req.url)
+        if allow_netrc and username is None and password is None:
+            netrc_auth = get_netrc_auth(original_url)
+            if netrc_auth:
+                logger.debug("Found credentials in netrc for %s", netloc)
             username, password = netrc_auth if netrc_auth else (None, None)
 
+        # If available, use keyring. If we have a username, use it, but
+        # allow keyring to provide one based on the URL.
+        if allow_keyring and password is None and keyring is not None:
+            try:
+                try:
+                    get_credential = keyring.get_credential
+                except AttributeError:
+                    logger.debug("Getting password from keyring for %s",
+                                 netloc)
+                    password = keyring.get_password(url, username)
+                    if password:
+                        logger.debug("Found password in keyring")
+                else:
+                    logger.debug("Getting credentials from keyring for %s",
+                                 netloc)
+                    cred = get_credential(url, username)
```

Comment:
Should `url` be `netloc`? It seems `url` is the url to the package, it seems silly to query user/pass per package, since the user/pass will typically be tied to the base url.

Also, it appears that you have to manually enter your credentials into the keyring prior to running the command. If the lookup fails to find your credentials, it should prompt you and then cache the credentials you entered into the keyring.

As, it is right now, users would need to modify the source code to print out the url that you're actually looking up the credentials with because its not obvious.

Reply:
Good point. `netloc` is certainly more generic, though there may be cases where the full URL is necessary (e.g. GitHub is a good example where you may have different credentials for more of the URL than just the netloc).

What if it tried `url` first and `netloc` as a fallback?

Reply:
To be clear, my expectation is that a full `url` would only be handled by keyring backends that do more than a simple store/lookup, and would trim the URL as far as they need. (For example, a specialized GitHub backend might negotiate a scoped access token and return that, rather than the all-access pass that a real username/password provides.)

Given this, I would be okay with only using `netloc` for calls to `get_password`, and only trying `url` when `get_credential` is supported (with a fallback to `netloc`).

Reply:
That would probably work. Our setup uses a Nexus pip repository and we only need to control access to the entire repository. I'm not sure if we can even setup permissions per package.

Are there plans to update the PR to also store credentials into the keystore with `set_password` after successful authentication?

Reply:
I like using the index URL.

Storing a typed password is an option, but potentially a surprising one. I'll ask up-level